### PR TITLE
Implement getters for WebIDL dictionaries

### LINKED ISSUE #1793: Add getters for WebIDL's dictionaries.
### Motivation

Currently `web-sys` wrappers do not have any getters for WebIDL dictionaries (e.g. [MediaStreamConstraints](https://docs.rs/web-sys/0.3.27/web_sys/struct.MediaStreamConstraints.html)). 

Such fields can be accessed using [js_sys::Reflect](https://docs.rs/js-sys/0.3.6/js_sys/struct.Reflect.html), which is quite awkward.

### Proposed Solution

Add getters for WebIDL dictionaries fields.

### Alternatives

Some generic getter like: `fn get(name:&str) -> JsValue`

### PR DESCRIPTION
**Pull request by pablosichert**
Heyo,

this PR implements getters for WebIDL dictionaries. The diff is bit noisy due to the autogenerated files; you can look at the actual changes to the code generation in the last two files `crates/webidl/src/generator.rs` and `crates/webidl/src/lib.rs` and corresponding tests in `crates/webidl-tests/dictionary.rs`.

Currently, setters are implemented as `my_property(&mut self, val: ...)`. To avoid breaking backwards compatibility, I have generated a new trait `MyDictionaryTypeGetters` for every `MyDictionaryType`. It contains definitions in the form of `my_property(&self) -> ...`.

Personally, I would suggest to rename all setters to `set_my_property` and place the `my_property` getters back in the `impl MyDictionaryType` as conventionally used in Rust. Preferably in another PR since as already mentioned would break backwards compatibility.

For the dictionary getter return types I have applied the same logic that is currently used for Interface getters:

https://github.com/rustwasm/wasm-bindgen/blob/e78db23553e70a48767a9bafaac3a4ffb4cba08d/crates/webidl/src/lib.rs#L688-L692

https://github.com/rustwasm/wasm-bindgen/blob/e78db23553e70a48767a9bafaac3a4ffb4cba08d/crates/webidl/src/generator.rs#L298-L300

This should close #1793 and #2921, and address point 1 of #3921.

### CODE REVIEW

File: crates/webidl/src/generator.rs [hunk 1]
```diff
@@ -815,6 +930,17 @@ impl Dictionary {
...
+            #unstable_attr
+            #getter_trait_doc_comment
+            pub trait #getter_trait_name {
+                #(#getter_trait_fields)*
+            }
+
+            #unstable_attr
+            impl #getter_trait_name for #name {
+                #(#getter_trait_impl_fields)*
+            }
```
Comment (pablosichert) [inline crates/webidl/src/generator.rs:942]: That's the main addition of this PR...

File: crates/webidl/src/generator.rs [hunk 2]
```diff
@@ -654,14 +656,97 @@ impl DictionaryField {
...
+        let Some(return_ty) = return_ty else {
+            return None;
+        };
+
+        quote! {
+            #unstable_attr
+            #cfg_features
+            #doc_comment
+            #unstable_docs
+            fn #name(&self) -> #return_ty;
```
Comment (pablosichert) [inline crates/webidl/src/generator.rs:713]: ...with getter signatures that are defined here...

File: crates/webidl/src/generator.rs [hunk 3]
```diff
@@ -654,14 +656,97 @@ impl DictionaryField {
...
+            return None;
+        };
+        let shim_name = self.shim_name();
+
+        quote! {
+            #unstable_attr
+            #cfg_features
+            fn #name(&self) -> #return_ty {
+                self.#shim_name()
+            }
```
Comment (pablosichert) [inline crates/webidl/src/generator.rs:744]: ...and implemented here

**Comment by MOZGIII** [comment]
> Personally, I would suggest to rename all setters to `set_my_property` and place the `my_property` getters back in the `impl MyDictionaryType` as conventionally used in Rust. Preferably in another PR since as already mentioned would break backwards compatibility.

I agree. We should bite the bullet, break the compatibility and do it (in another PR ofc).

**Comment by MOZGIII** [comment]
I like the idea of having traits here. We could probably go even further - have a trait with a WebIDL-generated declaration of the whole interface of a given type - not just for getters. That way we can include getters, setters and methods (both static and not).
The possible benefit from this would be nicer composability of wasm types.

I think we could create an issue do discuss this further, if this sounds like a viable direction to explore.

**Comment by pablosichert** [comment]
> The possible benefit from this would be nicer composability of wasm types.

I think you are referring to what I have an mind, but just to elaborate the thought: I'm not fully sure yet how hard it would be to make it possible, but if that would allow e.g. implementing `FooGetters` for `Bar` where `Bar: Foo`, that would be great. Currently you need to explicitly downcast the type.

File: crates/webidl-tests/dictionary.rs [hunk 4]
```diff
@@ -56,3 +56,30 @@ fn required() {
...
+        use crate::generated::ManyTypesGetters;
+        assert_eq!(many_types.a(), "a");
+        assert_eq!(many_types.n1(), 1);
+        assert_eq!(many_types.n2(), 2);
+        assert_eq!(many_types.n3(), 3);
+        assert_eq!(many_types.n4(), 4);
+        assert_eq!(many_types.n5(), 5);
+        assert_eq!(many_types.n6(), 6);
+        assert_eq!(many_types.c(), other_dict);
+    }
```
Comment (pablosichert) [inline crates/webidl-tests/dictionary.rs:84]: Unfortunately, due to the signature name clash this is a bit awkward to use right now.

If you move the `use crate::generated::ManyTypesGetters;` to the parent block, this test breaks.

**Comment by MOZGIII** [comment]
Let's take a [`WritableStream`](https://streams.spec.whatwg.org/#writablestream) for example.

It it was a trait, could just implement it for a `JsValue`, like so: `impl WritableStream for JsValue { ... }` (with some codegen of course).

Then, if needed, we could very elegantly extend `WritableStream` with [`WebTransportSendStream`](https://w3c.github.io/webtransport/#webtransportsendstream) with an explicit `trait WebTransportSendStream: WritableStream { ... }` - and also implement the `WebTransportSendStream` for `JsValue`.

---

<details><summary>WebIDL</summary>
<p>

```webidl
[Exposed=*, Transferable]
interface WritableStream {
  constructor(optional object underlyingSink, optional QueuingStrategy strategy = {});

  readonly attribute boolean locked;

  Promise<undefined> abort(optional any reason);
  Promise<undefined> close();
  WritableStreamDefaultWriter getWriter();
};
```

```webidl
[Exposed=(Window,Worker), SecureContext, Transferable]
interface WebTransportSendStream : WritableStream {
  attribute WebTransportSendGroup? sendGroup;
  attribute long long sendOrder;
  Promise<WebTransportSendStreamStats> getStats();
  WebTransportWriter getWriter();
};
```

</p>
</details> 

The traits would look somewhat like this:

<details><summary>Traits</summary>
<p>

```rust
trait WritableStream: Sized + JsAny {
    fn new() -> Self;
    fn new_with_underlying_sink(underlying_sink: &impl JsObject) -> Self;
    // ...

    fn locked(&self) -> bool;

    fn abort(&self) -> Promise<()>;
    fn abort_with_reason(&self, reason: &impl JsAny) -> Promise<()>;

    fn close(&self) -> Promise<()>;

    fn get_writer(&self) -> Result<impl WritableStreamDefaultWriter>;
}

trait WebTransportSendStream: WritableStream + JsAny {
    fn send_group(&self) -> Option<impl WebTransportSendGroup>;
    fn set_send_group(&self, value: &impl WebTransportSendGroup) -> Option<impl WebTransportSendGroup>;
    fn unset_send_group(&self);

    fn send_order(&self) -> i64;
    fn set_send_order(&self, value: i64);
    fn unset_send_order(&self);
    
    fn get_stats(&self) -> Promise<impl WebTransportSendStreamStats>;

    fn get_writer(&self) -> Result<impl WebTransportWriter>;
}
```

</p>
</details> 

and impls like this:


<details><summary>Impls</summary>
<p>

```rust
impl WritableStream for JsValue {
    fn new() -> Self {
        __shim_WritableStream_new()
    }

    fn new_with_underlying_sink(underlying_sink: &impl JsObject) -> Self {
        __shim_WritableStream_new_with_underlying_sink(underlying_sink)
    }

    // ... and so on.
}

impl WebTransportSendStream for JsValue {
    fn send_group(&self) -> Option<impl WebTransportSendGroup> {
        __shim_WebTransportSendStream_send_group()
    }
}
```

</p>
</details>

**Comment by pablosichert** [comment]
Cool, tests pass now.

Just a note that this will break for people that use `web_sys::*` in their code, but I guess they should be aware that they are opening up to that danger by using a wildcard.

**Comment by daxpedda** [comment]
I think that the easiest way forward would be to just implement getters as `get_my_property()`.

Obviously this isn't Rusty, but I believe its better then introducing a trait for each type while still remaining backwards compatible.

@Liamolucko would appreciate your input as well.

**Comment by Liamolucko** [comment]
> I think that the easiest way forward would be to just implement getters as `get_my_property()`.

I agree; I think the messiness of adding the extra traits outweighs the benefit of the better names.

I'm also not really comfortable with the potential breakage of `use web_sys::*`.

**Comment by pablosichert** [comment]
Do you keep track of changes that you'd want to include in a future breaking change? Since arguably this would be a great candidate, and the previous naming does not follow common Rust guidelines.

I can't imagine anyone complaining that their setters are now called `set_*` over of the current naming usually reserved to getters.

Even though I disagree with the naming, happy to make the necessary changes to `get_*` if this is what you'd prefer to merge in.

**Comment by daxpedda** [comment]
> Do you keep track of changes that you'd want to include in a future breaking change?

We don't, mainly because we don't see a breaking change on the horizon anytime soon.

> Even though I disagree with the naming, happy to make the necessary changes to `get_*` if this is what you'd prefer to merge in.

Indeed that's how we would like to proceed here.

**Comment by MOZGIII** [comment]
I was thinking about this, and there is a breaking change that I'd want to merge in with the async fs - that one didn't get any attention for a while not...

So, maybe it is time to fork? It would not be too bad to have an experimental fork, especially since most of the code is generated. I'd very much welcome the changes in here, and my use cases really really don't care about the breaking changes - as long as the underlying objects and be converted back and forth between the `JSValue` types of the fork and the upstream - that would be fine.

I can spend some resources setting up the repos and etc, but it would be great to collab with someone and formulate the needs for the fork upfront - like having more frequent breaking releases to allow advancement.

**Comment by daxpedda** [comment]
Replaced with #3993.

**Pull request closed without merging**